kotlinx-coroutines-core/jvm/test/scheduling/BlockingCoroutineDispatcherThreadLimitStressTest.kt

```diff
@@ -27,41 +27,36 @@ class BlockingCoroutineDispatcherThreadLimitStressTest : SchedulerTestBase() {
                 async(limitingDispatcher) {
                     try {
                         val currentlyExecuting = concurrentWorkers.incrementAndGet()
-                        observedConcurrency[currentlyExecuting] = true
-                        assertTrue(currentlyExecuting <= CORES_COUNT)
+                        observedParallelism.add(currentlyExecuting)
                     } finally {
                         concurrentWorkers.decrementAndGet()
                     }
                 }
             }
-            tasks.forEach { it.await() }
-            for (i in CORES_COUNT + 1..CORES_COUNT * 2) {
-                require(i !in observedConcurrency.keys) { "Unexpected state: $observedConcurrency" }
+            tasks.awaitAll()
+            assertTrue("Expected parallelism should be 1, had $observedParallelism") {
+                observedParallelism.single() == 1
             }
-            checkPoolThreadsCreated(0..CORES_COUNT + 1)
         }
     }
 
     @Test
-    @Ignore
     fun testLimitParallelism() = runBlocking {
         val limitingDispatcher = blockingDispatcher(CORES_COUNT)
         val iterations = 50_000 * stressTestMultiplier
         val tasks = (1..iterations).map {
             async(limitingDispatcher) {
                 try {
                     val currentlyExecuting = concurrentWorkers.incrementAndGet()
-                    observedConcurrency[currentlyExecuting] = true
-                    assertTrue(currentlyExecuting <= CORES_COUNT)
+                    observedParallelism.add(currentlyExecuting)
                 } finally {
                     concurrentWorkers.decrementAndGet()
                 }
             }
         }
-        tasks.forEach { it.await() }
-        for (i in CORES_COUNT + 1..CORES_COUNT * 2) {
-            require(i !in observedConcurrency.keys) { "Unexpected state: $observedConcurrency" }
+        tasks.awaitAll()
+        assertTrue("Unexpected state: $observedParallelism") {
+            observedParallelism.max() <= CORES_COUNT
         }
```

Comment:
Stings me every time as well.

However I would rather push our string interpolation towards laziness to get rid of lambdas where they are clearly redundant and exist "for performance"

Reply:
> I would rather push our string interpolation towards laziness

Please don't!

Reply:
Feel free to elaborate :)

Reply:
Tough to answer without an elaboration of how you expect to do that, but I'll try.

I understand the following options:

* (Compute the constituent strings eagerly, combine them later) You'd split the string construction cost into two parts: first, we'd build a list of all strings, then, on the call site, we'd combine them. For the fairly common use case of actually needing to use the string, you'd lose performance due to allocating the intermediate list; for the case where the string is not needed, you'd only save the `StringBuilder`'s calls, and are they really noticable compared to the string building inside `toString` calls of the components?
* (Postpone computing the constituent strings; do it safely) There's no notion of side-effect-free code in Kotlin, so the best we can hope to do *safely* is to introduce an unreliable do-it-later optimization that may or may not fire, making small tweaks to code result in outsized and unpredictable performance effects. Also, given how light on optimizations the Kotlin compiler is, I struggle to imagine anyone actually implementing this.
* (Postpone computing the constituent strings; do it universally) https://grep.app/search?q=%5C%24%7B%5B%5E%7D%5D%2B%5C%2B%5C%2B%7D&regexp=true&filter[lang][0]=Kotlin

Reply:
Thanks. Probably should've been more clear, I consider this idea to actual interpolation (the one we are trying to design), not what we have one (which is a glorified string concatenation)

Reply:
That's better if we can strictly limit string interpolation to immutable objects:
```
val log = mutableListOf<String>() // will contain essentially the same message many times
val queue = mutableListOf<Node>()
while (queue.isNotEmpty()) {
  log.add("{:?}".format(queue))
  queue.pop()
  // ...
}
```

Reply:
Would be nice, though there might be a little problem with the notion of immutability in the language 😅

Reply:
Since this is new API (that we fully control), we could do something like

```kotlin
interface Formattable {
  /** Return the string directly if your object is mutable, or face severe punishment. */
  fun format(directives: FormatDirectives): FormatResult
}

interface FormatResult

class String: FormatResult

class DeferredFormatResult: FormatResult
```